File: frontend/src/domain.ts
// Copia mínima del mapa de roles/módulos de app/domain.py (services/auth).
// La autorización real se aplica en el backend; esta copia solo controla
// qué se muestra en la navegación (E1-H2 escenario 1: "solo son visibles
// los módulos permitidos para ese rol").

export type Rol = "COMERCIAL" | "IMPORTACIONES" | "TECNICO" | "ADMINISTRATIVO" | "GERENCIA";

export type LineaNegocio = "MOBILITY" | "INDUSTRY" | "AMBAS";

export type EstadoUsuario = "ACTIVO" | "INACTIVO";

export type ModuleKey =
  | "comercial"
  | "reg-maestro"
  | "importaciones"
  | "tecnico"
  | "stock"
  | "financiero"
  | "logistica"
  | "administracion";

export const MODULE_LABELS: Record<ModuleKey, string> = {
  comercial: "Comercial",
  "reg-maestro": "Reg. maestro",
  importaciones: "Importaciones",
  tecnico: "Técnico",
  stock: "Stock",
  financiero: "Financiero",
  logistica: "Logística",
  administracion: "Administración",
};

export const ROLE_LABELS: Record<Rol, string> = {
  COMERCIAL: "Comercial",
  IMPORTACIONES: "Importaciones",
  TECNICO: "Técnico",
  ADMINISTRATIVO: "Administrativo",
  GERENCIA: "Gerencia",
};

export const LINE_LABELS: Record<LineaNegocio, string> = {
  MOBILITY: "Mobility",
  INDUSTRY: "Industry",
  AMBAS: "Ambas",
};

const DEDICATED_MODULE_ROUTES: Partial<Record<ModuleKey, string>> = {
  administracion: "/usuarios",
  "reg-maestro": "/reg-maestro",
  comercial: "/comercial",
  importaciones: "/importaciones",
  tecnico: "/tecnico",
  stock: "/stock",
  financiero: "/financiero",
  logistica: "/logistica",
};

export function moduleRoute(m: ModuleKey): string {
  return DEDICATED_MODULE_ROUTES[m] ?? `/modulos/${m}`;
}

// E1-H3: catálogos de la ficha central del proyecto (CRP).
// E5-H2: ENTREGADO es el estado final del proyecto (solo etapa_actual, no
// el estado por módulo).
export type EstadoEtapa = "PENDIENTE" | "EN_CURSO" | "CERRADO" | "BLOQUEADO" | "ENTREGADO";

export const ESTADO_ETAPA_LABELS: Record<EstadoEtapa, string> = {
  PENDIENTE: "Pendiente",
  EN_CURSO: "En curso",
  CERRADO: "Cerrado",
  BLOQUEADO: "Bloqueado",
  ENTREGADO: "Entregado",
};

export type SemaforoColor = "VERDE" | "AMARILLO" | "ROJO";

// E2-H1: catálogos del módulo Comercial (leads).
export type EstadoLead = "COTIZAR" | "ENVIADA" | "VENDIDO";

export const ESTADO_LEAD_LABELS: Record<EstadoLead, string> = {
  COTIZAR: "Cotizar",
  ENVIADA: "Enviada",
  VENDIDO: "Vendido",
};

export const CANAL_ENTRADA_OPTIONS = [
  "Sitio web",
  "Referido",
  "Feria comercial",
  "Redes sociales",
  "Llamada entrante",
] as const;

// E2-H2: catálogo de la calculadora oficial de cotizaciones.
export type TipoPago = "CONTADO" | "CREDITO";

export const TIPO_PAGO_LABELS: Record<TipoPago, string> = {
  CONTADO: "Contado",
  CREDITO: "Crédito",
};

// E2-H4: clasificación que se pide al marcar el lead como Vendido (E2-H3).
export type TipoClasificacion = "GM" | "STOCK_MOBILITY" | "STOCK_INDUSTRY";

export const TIPO_CLASIFICACION_LABELS: Record<TipoClasificacion, string> = {
  GM: "GM (Registro Maestro)",
  STOCK_MOBILITY: "Stock — Mobility",
  STOCK_INDUSTRY: "Stock — Industry",
};

// E4-H1: catálogo del checklist documental de importación.
export type TipoItemChecklist = "REQUERIDO" | "OPCIONAL";
export type EstadoItemChecklist = "PENDIENTE" | "ARCHIVADO" | "NO_APLICA";

// E5-H1/E5-H2: catálogo de la instalación (módulo Técnico).
export type EstadoInstalacion = "PROGRAMADO" | "COMPLETADO";

export const ESTADO_INSTALACION_LABELS: Record<EstadoInstalacion, string> = {
  PROGRAMADO: "Programado",
  COMPLETADO: "Completado",
};

export const ESTADO_ITEM_CHECKLIST_LABELS: Record<EstadoItemChecklist, string> = {
  PENDIENTE: "Pendiente",
  ARCHIVADO: "Archivado",
  NO_APLICA: "No aplica",
};

File: services/auth/app/domain.py
"""Catálogos de negocio (roles, líneas de negocio, estado, módulos).

Fuente única de verdad para las reglas de E1-H1 (gestión de usuarios) y
E1-H2 (control de acceso por módulo). El frontend mantiene una copia
mínima de ROLE_MODULE_ACCESS para filtrar la navegación; aquí es donde se
aplica realmente la autorización.
"""
from enum import Enum


class Rol(str, Enum):
    COMERCIAL = "COMERCIAL"
    IMPORTACIONES = "IMPORTACIONES"
    TECNICO = "TECNICO"
    ADMINISTRATIVO = "ADMINISTRATIVO"
    GERENCIA = "GERENCIA"


class LineaNegocio(str, Enum):
    MOBILITY = "MOBILITY"
    INDUSTRY = "INDUSTRY"
    AMBAS = "AMBAS"


class EstadoUsuario(str, Enum):
    ACTIVO = "ACTIVO"
    INACTIVO = "INACTIVO"


class Modulo(str, Enum):
    COMERCIAL = "comercial"
    REG_MAESTRO = "reg-maestro"
    IMPORTACIONES = "importaciones"
    TECNICO = "tecnico"
    STOCK = "stock"
    FINANCIERO = "financiero"
    LOGISTICA = "logistica"
    ADMINISTRACION = "administracion"


ROLE_MODULE_ACCESS: dict[Rol, set[Modulo]] = {
    Rol.GERENCIA: set(Modulo),
    # E6-H1: Comercial consulta la disponibilidad de stock (no la edita) al
    # clasificar una venta como Stock — Mobility/Industry.
    Rol.COMERCIAL: {Modulo.COMERCIAL, Modulo.REG_MAESTRO, Modulo.STOCK},
    Rol.IMPORTACIONES: {Modulo.IMPORTACIONES, Modulo.REG_MAESTRO},
    Rol.TECNICO: {Modulo.TECNICO, Modulo.STOCK, Modulo.REG_MAESTRO},
    # E6-H1/E8-H1: "Bodega"/"Logística" no son roles propios de la plataforma
    # (E1-H1 fija los roles); se mapean a Administrativo, que además ya
    # administra Financiero.
    Rol.ADMINISTRATIVO: {Modulo.FINANCIERO, Modulo.REG_MAESTRO, Modulo.STOCK, Modulo.LOGISTICA},
}


def can_manage_users(role: Rol) -> bool:
    """E1-H1: solo Gerencia/Administrador puede crear o modificar usuarios."""
    return role == Rol.GERENCIA


def can_access_module(role: Rol, module: Modulo) -> bool:
    """E1-H2 escenario 3: acceso a módulo no autorizado."""
    return module in ROLE_MODULE_ACCESS.get(role, set())


def accessible_modules(role: Rol) -> list[Modulo]:
    return sorted(ROLE_MODULE_ACCESS.get(role, set()), key=lambda m: m.value)
